### DISCUSSION TITLE
Suspense support in type definitions

### DISCUSSION THREAD

**Discussion by peterchappy** [TanStack/query#3470] [Q&A]
With React 18 out I've started looking into adopting `Suspense` more broadly across my application. I was wondering if there was any plan to model the types in a way that `UseQueryResult` is able account for when for the option `{ suspense: true }` so I don't need to still discriminate on `isSuccess` to keep the compiler happy when using suspense? 

I know that the suspense features are listed as experimental so I started working on a hack of a solution for now but obviously would prefer the types be baked in.

**Comment by yss14** [comment]
@peterchappy Do you think about something like this?

```typescript
const { data } = useQuery(['users'], () => getUsers(), { suspense: true }) // here type of <data> would be User[], and not User[] | undefined
```

If so, I also had the same thought, because at the moment we are also heavily using some self-written utilities (without suspense feature) to avoid repetitive patters like checking for `isSuccess` to make sure data is actually in place.

However, I think as outlined in https://github.com/tannerlinsley/react-query/issues/3310, there are still edge cases which would prevent omitting the `| undefined` from `data` in suspense mode. Problematic scenarios are cancelled queries during initial fetching or disabled queries. In such cases `data` could still be `undefined` even when using suspense mode.

**Comment by TkDodo** [comment]
> However, I think as outlined in https://github.com/tannerlinsley/react-query/issues/3310, there are still edge cases which would prevent omitting the | undefined from data in suspense mode. Problematic scenarios are cancelled queries during initial fetching or disabled queries. In such cases data could still be undefined even when using suspense mode.

exactly, a simple:

```
const { data } = useQuery(['users'], () => getUsers(), { suspense: true, enabled: false })
```

would give you `undefined` until the query gets enabled. So I'm afraid it's not that easy.

**Reply by joshribakoff-sm** [reply]
Couldn't it be done with generics and conditional types, so the return type depends on the values passed as arguments?

**Reply by yss14** [reply]
@joshribakoff-sm I think some cases might be solved via generics and conditional types, but for example the case where a query is cancelled during the initial fetching can unfortunately not be solved via generics.
